### ISSUE TITLE
Airflow2 : When clearing an parent subdag, all child subdags are not cleared

### ISSUE THREAD

**Issue by keomgak** [apache/airflow#31491]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

I'm using subdag in airflow2. My subdags is like this,,

```
parent_subdag_sample
\child_subdag_start (success)
\child_subdag_task_1 (success)
\child_subdag_task_2 (success)
\child_subdag_task_fail (fail)
\child_subdag_task_3 (upstream_failed)
\child_subdag_end (upstream_failed)
```
The case when child_subdag_task_fail is failed,, I want to rerun from the first task(child_subdag_start). So, I clear parent_subdag_sample with the recursive + downstream option selected. (I did not select the failed option)

For airflow1, from child_subdag_start task is cleared. For airflow2.4.3, child_subdag_start, child_subdag_task_1, and child_subdag_task_2 with success status are not cleared, but from child_subdag_task_fail task is cleared.

The recursive option is described as "Recursive - All the tasks in the child DAGs and parent DAGs

So I think all child subdag will be cleared like airflow1.
In airflow2.4.3, I wonder why child subdag that succeeded is not cleared. 
Is there any way to make all child subdag clear?

### What you think should happen instead

When clearing an parent subdag, all child subdags should be cleared.

### How to reproduce

Here is my testcode.

```
from time import sleep
from pendulum.tz.timezone import Timezone
from datetime import datetime, timedelta
from airflow import DAG
from airflow.operators.python import PythonOperator
from airflow.operators.subdag import SubDagOperator
import subdag.child_subdag as cs

kst=Timezone('Asia/Seoul')

default_args={
    'owner': 'test',
    'retries':0,
    'retry_delay':timedelta(minutes=1),
}

def dump() -> None:
    sleep(3)


dag_id = 'subdag_test'

with DAG(
    dag_id=dag_id,
    default_args=default_args,
    start_date=datetime(2023,3,11, tzinfo=kst),
    end_date=datetime(2023,3,11, tzinfo=kst),
    schedule_interval="@daily",
    tags=['test', 'subDAG']
) as dag:
   parent_task_start = PythonOperator(
        task_id='parent_task_start',
        python_callable=dump,
    )

    parent_subdag_sample = SubDagOperator(
        task_id='parent_subdag_sample',
        subdag=cs.child_subdag(parent_dag_name=dag_id, child_dag_name="parent_subdag_sample")
    )

    parent_task_end= PythonOperator(
        task_id='parent_task_end',
        python_callable=dump,
    )

    parent_task_start >> parent_subdag_sample >> parent_task_end
```

```
def dump_error() -> None:
    raise AirflowFailException("fail")


default_args={
    'owner': 'test',
    'retries':0,
    'retry_delay':timedelta(minutes=1),
}

def child_subdag(parent_dag_name, child_dag_name) -> DAG:
    with DAG(
        dag_id=f"{parent_dag_name}.{child_dag_name}",
        default_args=default_args,
        schedule_interval="@daily",
        start_date=datetime(2023,3,11, tzinfo=kst),
        end_date=datetime(2023,3,11, tzinfo=kst),
        tags=['test', 'subdag', 'sample1'],
    ) as dag:
        child_subdag_start = PythonOperator(
            task_id='child_subdag_start',
            python_callable=dump
        )

        child_subdag_task_1 = PythonOperator(
            task_id='child_subdag_task_1',
            python_callable=dump
        )

        child_subdag_task_2 = PythonOperator(
            task_id='child_subdag_task_2',
            python_callable=dump
        )

        child_subdag_task_fail = PythonOperator(
            task_id='child_subdag_task_fail',
            python_callable=dump_error
        )

        child_subdag_task_3 = PythonOperator(
            task_id='child_subdag_task_3',
            python_callable=dump
        )

        child_subdag_end = DummyOperator(
            task_id='child_subdag_end'
        )

        child_subdag_start >> child_subdag_task_1 >> child_subdag_task_2 >> child_subdag_task_fail \
        >> child_subdag_task_3 >> child_subdag_end

        return dag
```

### Operating System

ubuntu 22.04

### Versions of Apache Airflow Providers

composer-2.1.8
airflow-2.4.3

### Deployment

Google Cloud Composer

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
Can you use the inspect option on your browser and check what is the payload sent to the backend when you click the clear button?

**Comment by keomgak** [comment]
> Can you use the inspect option on your browser and check what is the payload sent to the backend when you click the clear button?

Here is payload text.

clear(first)
```
csrf_token: IjA3NjhhMmE1OWRlNTI1YzAwMTY4NGExMWJiZjA3MmM0OGJiODlhYjYi.ZG6oYw.IlSPJfrsAQOy5r9sk3ZX0zQM8Ys
dag_id: test_main.test-child
dag_run_id: scheduled__2023-04-01T23:00:00+00:00
task_id: test-grandchild
confirmed: false
execution_date: 2023-04-01T23:00:00+00:00
past: false
future: false
upstream: false
downstream: true
recursive: true
only_failed: false
```

clear(second)
```
csrf_token: IjA3NjhhMmE1OWRlNTI1YzAwMTY4NGExMWJiZjA3MmM0OGJiODlhYjYi.ZG6oYw.IlSPJfrsAQOy5r9sk3ZX0zQM8Ys
dag_id: test_main.test-child
dag_run_id: scheduled__2023-04-01T23:00:00+00:00
task_id: test-grandchild
confirmed: true
execution_date: 2023-04-01T23:00:00+00:00
past: false
future: false
upstream: false
downstream: true
recursive: true
only_failed: false
```

**Comment by eladkal** [comment]
@hussein-awala if this affect only SubDAGs I would consider this as won't fix.
SubDags are deprecated for a long time. If this is a simple fix that is OK but if not I'm not sure if fixing deprecated feature worth the risk of breaking something else.

Users are asked to migrate to TaskGroups.

**Comment by hussein-awala** [comment]
@eladkal Actually the fix is simple and ready, I was going to ask you if it's acceptable or not to fix a deprecated operator, for the new feature I know that it is not, but what about fixing a bug? (knowing that the fix is only related to the operator itself and nothing else).

**Comment by hussein-awala** [comment]
@uranusjr In Airflow 1.x, the sub dag operator was completely different, it was based on backfill instead of scheduler, and I believe we used to clear the sub dag run when clearing the sub dag TIs. However, this behavior was modified in the [this PR](https://github.com/apache/airflow/commit/0be39219cd058ba7d50cdf34b2cc46513f4f5ab3) , which introduced the logic to only clear the failed TIs.
> This does not sound like something that’d change with refactoring and I wonder if “fixing” this would break someone that considered the 1.x behaviour as a bug.

It appears that it was indeed something changed during the refactoring.

**Comment by keomgak** [comment]
Dear all,

I know that subdag is deprecated. But for some reason, there are people like me who haven't converted to taskgroups yet. This is important to me. I would like you to review the bugfix.

**Comment by potiuk** [comment]
I would be agains fixing it. I think the milk has spilled a long time ago and reverting the behaviour after it has been changed and released in 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6 will bring more harm than good. If anything this bring more harm than good. 

I believe we already  have good solution for that in task group https://github.com/apache/airflow/pull/26658  (available to 2.5) and if anything we should not fix it, but we should use it drive people who want to use it to switch to Task Group.

> I know that subdag is deprecated. But for some reason, there are people like me who haven't converted to taskgroups yet. This is important to me. I would like you to review the bugfix.

Yes. And  - to be perfectly honest and brutal - by deprecating and providing the replacement, we want to drive them out of using subdag as soon as possible. So while not intentional, the result of this bug is that people like you are more likely to have incentive to switch to the Task Group.

Also - look at it from the other perspective. You probably don't realise that we would only implement such change in main and release in 2.7 if we decide to do so, So maybe it will get relased in a few months and it will get few more weeks to get 2.7.1 tha will have some teething problems solved.. Then it will take few more months for composer to support that version - and only then you will have to migrate going through all the hoops of going 2.4 -> 2.7 and ony then you wil be able to use the fix, providing that it works. You are looking at maybe 6 months + migration between 3 versions of airlfow to use the fix you propose.

Alternative - you migrate now to 2.5 and switch to Task Group. 

Thinks which way is more costly to your business.

**Comment by hussein-awala** [comment]
@keomgak As explained by @potiuk, this operator is end-of-development (EOD), so you should switch to Task Group as soon as possible.
However, if you are unable to do so at the moment and need a temporary solution, you can extend the operator and override the method `_reset_dag_run_and_task_instances` just like I did in #31545. You can use this custom operator until you migrate to Task Group.

I close the issue as won't fix.
